ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/bytes/VarInt.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("NOTHING_TO_INLINE")
+
+package io.ktor.network.quic.bytes
+
+import io.ktor.network.quic.util.*
+import io.ktor.utils.io.core.*
+
+internal inline fun ByteReadPacket.readVarIntOrElse(elseBlock: () -> Long): Long {
+    val value = readVarInt()
+    if (value == -1L) {
+        elseBlock()
+    }
+    return value
+}
+
+/**
+ * Reads variable-length non-negative integer value. Returns null if EOF.
+ *
+ * [RFC Reference](https://www.rfc-editor.org/rfc/rfc9000.html#name-variable-length-integer-enc)
+ */
+@OptIn(ExperimentalUnsignedTypes::class)
+private fun ByteReadPacket.readVarInt(): Long {
+    val value = readUByteOrElse { return -1 }.toInt()
+    val length = 1 shl (value ushr 6)
+
+    var varInt = (value and 0x3f).toLong()
+    var i = 0
+    while (i < length - 1) {
```

Comment:
may be we can unroll this loop here? It will be then symmetric to `writeVarInt` and again, easier to read/understand :)

Reply:
Actually this one from specification https://www.rfc-editor.org/rfc/rfc9000.html#name-sample-variable-length-inte

Reply:
Yeah, I see it now
But, IMO, It looks better (easier to debug/understand) without loop
Look for example on `quiche` impl - https://github.com/cloudflare/quiche/blob/1aa5c702332bf0c90984f5470e8cb750d2dd027f/octets/src/lib.rs#L406-L469